src/java/org/apache/cassandra/security/ISslContextFactory.java

```diff
@@ -0,0 +1,118 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.security;
+
+import java.util.List;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLException;
+
+import io.netty.handler.ssl.CipherSuiteFilter;
+import io.netty.handler.ssl.SslContext;
+
+/**
+ * The purpose of this interface is to provide pluggable mechanism for creating custom JSSE and Netty SSLContext
+ * objects. Please use the Cassandra configuration key {@code ssl_context_factory} as part of {@code
+ * client_encryption_options}/{@code server_encryption_options} and provide a custom class-name implementing this
+ * interface with parameters to be used to plugin a your own way to load the SSLContext.
+ *
+ * Implementation of this interface must have a constructor with argument of type {@code Map<String,Object>} to allow
+ * custom parameters, needed by the implementation, to be passed from the yaml configuration. Common SSL
+ * configurations like {@code protocol, algorithm, cipher_suites, accepted_protocols, require_client_auth,
+ * require_endpoint_verification, enabled, optional} will also be passed to that map by Cassanddra.
+ *
+ * Since on top of Netty, Cassandra is internally using JSSE SSLContext also for certain use-cases- this interface
+ * has methods for both.
+ *
+ * Below is an example of how to configure a custom implementation with parameters
+ * <pre>
+ * ssl_context_factory:
+ *       class_name: org.apache.cassandra.security.YourSslContextFactoryImpl
+ *       parameters:
+ *         key1: "value1"
+ *         key2: "value2"
+ *         key3: "value3"
+ * </pre>
+ */
+public interface ISslContextFactory
+{
+    /**
+     * Creates JSSE SSLContext.
+     *
+     * @param buildTruststore {@code true} if the caller requires Truststore; {@code false} otherwise
+     * @return JSSE's {@link SSLContext}
+     * @throws SSLException in case the Ssl Context creation fails for some reason
+     */
+    SSLContext createJSSESslContext(boolean buildTruststore) throws SSLException;
+
+    /**
+     * Creates Netty's SslContext object.
+     *
+     * @param buildTruststore {@code true} if the caller requires Truststore; {@code false} otherwise
+     * @param socketType {@link SocketType} for Netty's Inbound or Outbound channels
+     * @param useOpenSsl {@code true} if openSsl is enabled;{@code false} otherwise
```

Comment:
I think I get your suggestion after looking at how it is used but I need little bit more time to think over it.

Reply:
On the comment around triggering the hot reloading, I feel having a reasonable polling check from the Cassandra could be okay as far as the polling period is not very long. So the implementation can have aggressive check to know if something changed about keys/certs, but it can still wait couple of more seconds may be before applying the change. What do you think?

Reply:
I think that would probably work. There's a very small time-of-check, time-of-use issue that I'm not sure if it would be a problem in that you say 'ready to reload' and 'reload'. CEP-9 was considering the use case where the key/trust store materials are retrieved from some remote service.  It's probably ok to poll for changes and load, and probably just as bad as the local filesystem as you can't atomically replace both the keystone and truststore at the same time if they were separate files.

If the implementation does have to track some kind of state about whether to reload, when should that be cleared? Is there a contract with the interface that if `shouldReload()` is called the SSLContext will be revalidated and the cache cleared?

So the other thing that could be an issue is configurations where different SSL factories are used for the client and internode connections.  At the moment if either of the configurations flags a reload, then both are reloaded which may be unexpected by an implementation.

I think it would be cleaner to move all of the hot-loading stuff moved out to something like `HotReloadingSslFactory` and just expose an `SSLFactory.validateAndRefreshCache` method that can be called to indicate the cache is now stale and can be cleared.

The 'cache' itself is a bit weird, only likely containing at most eight items, and more likely just two.

Reply:
I've to still think about this bit more. 

Three points to still address (apart from providing an example),
1. I agree that currently - as soon as either of the configurations (server or client) changes, it triggers reloading for both which is unnecessary. 
2. I've to address the SSLFactory.validateAndRefreshCache related comment and 
3. `useOpenSsl` flag's removal.

Reply:
@jonmeredith Out of the above three - I moved the useOpenSsl flag determination to the AbstractSslContextFactory implementation. Please review. Now will work on the other two.

Reply:
Hi @jonmeredith can you please take a look at this last set of comments here? Beyond this I guess we are all done with code review comments. I'll update contract in the ISslContextFactory interface and ask for review and wait for the documentation guidelines in docs/ folder.

Reply:
Sorry I missed responding here. I think this is the last thing to resolve, everything else looked good to me.

I think invalidating for the appropriate option is just fine.  I think your `shouldReload()` changes make sense, but will need to see the final commit to be sure.

Reply:
Thanks @jonmeredith You can see the changes done for separating hot reloading trigger for client and server [here](https://github.com/maulin-vasavada/cassandra/pull/1/files). I can add that to this PR for the further review once you can take a look at it.

Reply:
Added some comments on your self-PR :)

Reply:
@jonmeredith I've made the changes on this PR based on the review in the self-PR :) Btw, thanks for that review before I could bring it here.